### PR TITLE
Feat: Add custom function framework

### PR DESCRIPTION
**Pull request by newborn22**
## Related Issue(s) & Descriptions
#517 

## Checklist

-   [x] Tests were added or are not required
-   [ ] Documentation was added or is not required

### CODE REVIEW

File: go/vt/vtgate/plan_execute.go [hunk 1]
```diff
@@ -508,3 +518,47 @@ func HandleWescaleFilterRequest(stmt sqlparser.Statement, vcursor *vcursorImpl,
...
+	c.Input = plan.Instructions
+	err := c.SetSentExprs(stmt)
+	if err != nil {
+		return nil, err
+	}
+
+	planForCustomFunction := &engine.Plan{}
+	*planForCustomFunction = *plan
+	planForCustomFunction.Instructions = c
+
```
Comment (earayu) [inline go/vt/vtgate/plan_execute.go:562]: I think planForCustomFunction needs a deep copy.
`BindVarNeeds` and `Warnings` are shadow copies now.

File: go/vt/vtgate/engine/projection.go [hunk 2]
```diff
@@ -33,6 +34,12 @@ type Projection struct {
 	Exprs []evalengine.Expr
 	Input Primitive
 	noTxNeeded
+	IsCustomFunctionProjection bool
+	Meta                       *CustomFunctionProjectionMeta
+}
+
+type CustomFunctionProjectionMeta struct {
+	Origin sqlparser.SelectExprs
```
Comment (earayu) [inline go/vt/vtgate/engine/projection.go:42]: Are these code not needed anymore?

Comment (newborn22) [reply go/vt/vtgate/engine/projection.go]: already removed

File: go/vt/vtgate/plan_execute.go (hunk 1 above)
Comment (newborn22) [reply go/vt/vtgate/plan_execute.go:562]: `BindVarNeeds` and `Warnings` are shadow copies so that changes to these variables can be applied on plan cache.

**Pull request merged**